pkg/nodediscovery/nodediscovery.go

```diff
@@ -454,6 +456,28 @@ func (n *NodeDiscovery) mutateNodeResource(nodeResource *ciliumv2.CiliumNode) er
 	// CiliumNode object
 	nodeResource.ObjectMeta.Annotations = k8sNodeParsed.GetCiliumAnnotations()
 
+	// This function can be called before we have restored the CiliumInternalIP.
+	// In that case, we do not want to remove the old CiliumInternalIP, as this
+	// would lead to the IP address flapping. Therefore, this code preserves any
+	// CiliumInternalIP if (and only if) the local node store does not yet
+	// include the restored CiliumInternalIP.
+	nodeResource.Spec.Addresses = slices.DeleteFunc(nodeResource.Spec.Addresses, func(address ciliumv2.NodeAddress) bool {
+		if address.Type == addressing.NodeCiliumInternalIP {
+			// Only delete a CiliumInternalIP if
+			// a) its IP family is disabled,
+			// and/or
+			// b) the LocalNode store contains an IP address which we can use instead
+			switch net.IPFamilyOfString(address.IP) {
+			case net.IPv4:
+				return !n.LocalConfig.EnableIPv4 || n.localNode.GetCiliumInternalIP(false) != nil
```

Comment:
So on v1.14 and older the behavior here is slightly different than on v1.15+: `n.localNode` is not the source of truth, rather it is populated via `n.fillLocalNode` (which is not always called before this function).

However, looking at the fact that we're still using `n.localNode.IPAddresses` to populate `nodeResource.Spec.Addresses` I think this is still correct. The important thing is that both pieces of code populate using the same source.